src/python/pants/engine/README.md

```diff
@@ -76,52 +76,72 @@ In cases where this search detects any ambiguity (generally because there are tw
 fail with a useful error message.
 
-### Datatypes
+### How to Declare Engine `Param` Types
 
 In practical use, builtin types like `str` or `int` do not provide enough information to
-disambiguate between various types of data in `@rule` signatures, so declaring small `datatype` 
-definitions to provide a unique and descriptive type is highly recommended:
+disambiguate between various types of data in `@rule` signatures, so declaring small, unique classes
+to encapsulate different states can help to make `@rule` sets with complex control flow between rules
+more declarative and self-documenting.
+
+#### Requirements for an Engine `Param`
+
+To use an instance of some class `C` as a `Param` to the engine:
+1. instances of `C` must be immutable,
+2. `__hash__` and `__eq__` must be implemented such that when constructing two separate instances of `C` with the same argument list `...`, `C(...) == C(...)` and `hash(C(...)) == hash(C(...))`.
+  - This can be ignored for singleton `Param`s.
+
+#### Benefits of using `@dataclass` for a `Param`
+[Python 3 `@dataclass`es](https://docs.python.org/3/library/dataclasses.html) satisfy the above requirements for engine `Param`s. Using `@dataclass` to declare engine `Param` types also provides additional benefits compared to alternatives:
+1. a compact and high-performance representation which can be stably shared across FFI boundaries,
+2. static type checking via [the `dataclasses` mypy plugin](https://github.com/python/mypy/blob/master/mypy/plugins/dataclasses.py),
+3. a concise, standard, and Pythonic way to declare classes.
+
+
+#### Example Usage of `@dataclass` for Engine `Param`s
+
+*Note that the [3rdparty `dataclasses` library](https://github.com/ericvsmith/dataclasses) must be in your BUILD file and `import`ed if you're running Pants with Python 3.6!*
 
 ```python
-class FormattedInt(datatype(['content'])): pass
+from dataclasses import dataclass
 
-@rule
-def int_to_str(value: int) -> FormattedInt:
-  return FormattedInt('{}'.format(value))
+from pants.util.meta import frozen_after_init
 
-# Field values can be specified with positional and/or keyword arguments in the constructor:
-x = FormattedInt('a string')
-x = FormattedInt(content='a string')
+# Pants requires that engine Params have a stable hash. This can be accomplished with the
+# `frozen=True` argument set in the `@dataclass` decorator.
+@dataclass(frozen=True)
+class FormattedInt:
+  content: str
 
-# Field values can be accessed after construction by name or index:
-print(x.content)    # 'a string'
-print(x[0])         # 'a string'
+# `@frozen_after_init` can also be used with `unsafe_hash=True` to create engine Params which can
+# modify their fields within the `__init__()` method:
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class ValidatedInt:
+  x: int
 
-# datatype objects can be easily inspected:
-print(x)            # 'FormattedInt(content=a string)'
-```
+  def __init__(self, x):
+    self.x = x
```

Comment:
I recommend some tweaks to explain why you would do this. Likewise, split this into a dedicated code block separate from the above one and move the description into normal text.

> `@frozen_after_init` can also be used with `unsafe_hash=True` to create engine Params which can modify their fields within the `__init__()` method. This is useful if you want to perform any normalization on the args to `__init__()`, such as allowing `List` and then casting this to an immutable `Tuple` in `__init__()`:

```python
from typing import Iterable, Tuple
from pants.util.meta import frozen_after_init

@frozen_after_init
@dataclass(unsafe_hash=True)
class ValidatedCollection:
  x: Tuple[str, ...]   # Note that we use a `Tuple` because `List` is mutable

  def __init__(self, x: Iterable[str]) -> None:
    self.x = tuple(x)
```

Reply:
Thank you so much for the thoughtful and detailed writing here! Will do!